secondary_structure/dot_bracket.go

```diff
@@ -0,0 +1,533 @@
+package secondary_structure
+
+import (
+	"fmt"
+	"regexp"
+)
+
+/******************************************************************************
+July, 01, 2021
+
+This file defines the structs and functions needed to get the 'annotated'
+structure and `SecondaryStructure` of a RNA from its 'dot-bracket' notation.
+
+'Dot-bracket' notation of a secondary structure is a string where each
+character represents a base. Unpaired nucleotides are represented with a '.'
+and base pairs are represented by paranthesis. '(' denotes the opening base
+and ')' denotes the closing base of a base pair.
+For example, "..((..)).." denotes a hairpin where the bases at index 2 and 7,
+and at index 3 and 6 are paired.
+
+'Annotated' structure of a secondary structure is a string where each
+character represents a base. In this notation, unpaired nucleotides are set
+to a character depending on what part of the secondary structure is in.
+The character mapping for the unpaired nucleotides is:
+* On the exterior part of a RNA structure (not part of any loop): 'e'
+* On the single stranded region of a hairpin: 'h'
+* On a single stranded region of a multiloop: 'm'
+* Part of an interior loop in a stem of a hairpin or multiloop: 'i'
+For example, "..((..)).." would have an annotated structure of "ee((hh))ee".
+Note that the paranthesis surrounding one or more 'h' or 'm' characters form
+the stem of that hairpin or multiloop respectively. Thus, in the example above,
+the bases at index 2, 3, 6, and 7 form the stem of the hairpin loop enclosed
+by the bases at index 3 and 6.
+
+ ******************************************************************************/
+
+const (
+	dotBracketUnpairedNucleotide       byte = '.'
+	dotBracketStemFivePrimeNucleotide  byte = '('
+	dotBracketStemThreePrimeNucleotide byte = ')'
+
+	exteriorLoopUnpairedNucleotide          byte = 'e'
+	interiorLoopUnpairedNucleotide          byte = 'i'
+	hairpinLoopNucleotide                   byte = 'h'
+	multiLoopSingleStrandedRegionNucleotide byte = 'm'
+)
+
+// parseCompound holds all information needed to compute the annotated structure
+// and `SecondaryStructure` of a RNA sequence
+type parseCompound struct {
+	length             int   // length of `sequence`
+	pairTable          []int // (see `pairTable()`)
+	annotatedStructure []byte
+}
+
+// FromDotBracket returns the annotated structure and `SecondaryStructure` of
+// a RNA sequence from its 'dot-bracket' structure.
+func FromDotBracket(dotBracketStructure string) (string, *SecondaryStructure, error) {
```

Comment:
The exported function, FromDotBracket, returns a string, a secondary structure, and an error. In the tests, I see only the string output is used. Other than for testing, why is *SecondaryStructure returned?